components/core/src/clp/streaming_compression/zstd/Decompressor.cpp

```diff
@@ -34,67 +34,33 @@ auto Decompressor::try_read(char* buf, size_t num_bytes_to_read, size_t& num_byt
 
     num_bytes_read = 0;
 
-    ZSTD_outBuffer decompressed_stream_block = {buf, num_bytes_to_read, 0};
+    ZSTD_outBuffer decompressed_stream_block{buf, num_bytes_to_read, 0};
     while (decompressed_stream_block.pos < num_bytes_to_read) {
         // Check if there's data that can be decompressed
-        if (m_compressed_stream_block.pos == m_compressed_stream_block.size) {
-            switch (m_input_type) {
-                case InputType::CompressedDataBuf:
-                    // Fall through
-                case InputType::MemoryMappedCompressedFile:
-                    num_bytes_read = decompressed_stream_block.pos;
-                    if (0 == decompressed_stream_block.pos) {
-                        return ErrorCode_EndOfFile;
-                    } else {
-                        return ErrorCode_Success;
-                    }
-                    break;
-                case InputType::ReaderInterface: {
-                    if (false == m_read_buffer.has_value()) {
-                        throw OperationFailed(ErrorCode_Corrupt, __FILENAME__, __LINE__);
-                    }
-                    auto& read_buffer{m_read_buffer.value()};
-                    auto error_code = m_reader->try_read(
-                            read_buffer.data(),
-                            read_buffer.size(),
-                            m_read_buffer_length
-                    );
-                    if (ErrorCode_Success != error_code) {
-                        if (ErrorCode_EndOfFile == error_code) {
-                            num_bytes_read = decompressed_stream_block.pos;
-                            if (0 == decompressed_stream_block.pos) {
-                                return ErrorCode_EndOfFile;
-                            } else {
-                                return ErrorCode_Success;
-                            }
-                        } else {
-                            return error_code;
-                        }
-                    }
-
-                    m_compressed_stream_block.pos = 0;
-                    m_compressed_stream_block.size = m_read_buffer_length;
-                    break;
-                }
-                default:
-                    throw OperationFailed(ErrorCode_Unsupported, __FILENAME__, __LINE__);
+        if (m_compressed_stream_block.pos == m_compressed_stream_block.size
+            && false == m_frame_has_more_data)
+        {
+            auto const error_code{refill_compressed_buffer()};
+            if (ErrorCode_Success != error_code) {
+                num_bytes_read = decompressed_stream_block.pos;
+                return decompressed_stream_block.pos > 0 ? ErrorCode_Success : error_code;
```

Comment:
I think another bug in the existing code is that we don't update `m_decompressed_stream_pos` before we return here. Can you add it?